pkg/credentialprovider/plugin/config.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugin
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/criauth"
+)
+
+// readCRIAuthPluginConfigFile receives a path to a config file and decodes it
+// into the internal CRIAuthPluginConfig type.
+func readCRIAuthPluginConfigFile(configPath string) (*criauth.CRIAuthPluginConfig, error) {
+	if configPath == "" {
+		return nil, fmt.Errorf("CRI auth plugin config path is empty")
+	}
+
+	data, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read external registry credential provider configuration from %q [%v]", configPath, err)
+	}
+
+	return decode(data)
+}
+
+// decode decodes data into the internal CRIAuthPluginConfig type.
+func decode(data []byte) (*criauth.CRIAuthPluginConfig, error) {
+	obj, gvk, err := codecs.UniversalDecoder().Decode(data, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if gvk.Kind != "CRIAuthPluginConfig" {
+		return nil, fmt.Errorf("failed to decode %q (missing Kind)", gvk.Kind)
+	}
+
+	if internalConfig, ok := obj.(*criauth.CRIAuthPluginConfig); ok {
+		return internalConfig, nil
+	}
+
+	return nil, fmt.Errorf("unable to convert %T to *CRIAuthPluginConfig", obj)
+}
+
+func validateCRIAuthPluginConfig(config *criauth.CRIAuthPluginConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(config.Plugins) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("plugins"), "at least 1 item in plugins is required"))
+	}
+
+	fieldPath := field.NewPath("plugins")
+	for _, plugin := range config.Plugins {
+		if len(plugin.ImageMatchURLs) == 0 {
+			allErrs = append(allErrs, field.Required(fieldPath.Child("matchURLs"), "at least 1 item in matchUrls is required"))
+		}
+
+		if plugin.DefaultCacheDuration.Duration == 0 {
```

Comment:
The plugin response can say "don't cache this response", but I think the plugin config should always indicate some default caching duration in the event the response has a nil cache duration.

Reply:
Curious, why 0 duration (no cache) cannot be an option? If this is to fool proof configuration to prevent users from doing no cache plugins, should the minimal timeout be specified?

Maybe you meant condition `if plugin.DefaultCacheDuration == nil {` here?

Reply:
> Curious, why 0 duration (no cache) cannot be an option? If this is to fool proof configuration to prevent users from doing no cache plugins, should the minimal timeout be specified?

0 cache duration is an option but in CRIAuthPluginResponse. This is validating the `defaultCacheDuration` config which is the fallback cache duration if the plugin didn't specify one.